rust/src/storage/file.rs

```diff
@@ -58,26 +65,110 @@ impl StorageBackend for FileStorageBackend {
     }
 
     async fn put_obj(&self, path: &str, obj_bytes: &[u8]) -> Result<(), StorageError> {
-        let dir = std::path::Path::new(path);
-
-        if let Some(d) = dir.parent() {
-            fs::create_dir_all(d).await?;
-        }
+        let tmp_file = self
+            .root
+            .join("_tmp_writes")
+            .join(Uuid::new_v4().to_string());
 
-        // use `OpenOptions` with `create_new` to create the file handle.
-        // this will force ErrorKind::AlreadyExists if the file already exists.
-        let mut f = fs::OpenOptions::new()
-            .create_new(true)
-            .write(true)
-            .open(path)
-            .await
-            .map_err(|e| match e.kind() {
-                std::io::ErrorKind::AlreadyExists => StorageError::AlreadyExists(path.to_string()),
-                _ => StorageError::Io { source: e },
-            })?;
+        create_parent_dirs(&tmp_file).await?;
+        create_parent_dirs(Path::new(path)).await?;
```

Comment:
Not a big fan of calling this on each time. I'm thinking of moving `put_obj` into helper function and use a wrapper with error handling to catch missing/uncreated parent dirs and create them once and when needed instead of calling create_parent_dirs each time. Wdyt?

Reply:
yeap, make sense